### PR TITLE
Adding support to link a tld to a child IBDO for inheriting parameters

### PR DESCRIPTION
**Pull request by rpg36**
An effort to link children to the top level document to inherit their metadata instead of copying it down.

### CODE REVIEW

File: src/main/java/emissary/core/BaseDataObject.java [hunk 1]
```diff
@@ -194,6 +194,10 @@ public class BaseDataObject implements Serializable, Cloneable, Remote, IBaseDat
 
     final SafeUsageChecker safeUsageChecker = new SafeUsageChecker();
 
+    protected final IBaseDataObject parent;
+
+    protected final Set<String> inheritedKeys;
```
Comment (fbruton) [inline src/main/java/emissary/core/BaseDataObject.java:199]: Recommend removing this field from the BaseDataObject` and passing as a parameter to `IBaseDataObjectHelper#getParametersWithInheritance()`

File: src/main/java/emissary/core/BaseDataObject.java [hunk 2]
```diff
@@ -194,6 +194,10 @@ public class BaseDataObject implements Serializable, Cloneable, Remote, IBaseDat
 
     final SafeUsageChecker safeUsageChecker = new SafeUsageChecker();
 
+    protected final IBaseDataObject parent;
```
[the remark this answers is not in the document]
Comment (fbruton) [reply src/main/java/emissary/core/BaseDataObject.java:197]: I agree with @drivenflywheel. Parent and TLD have different meanings when the BaseDataObject is more than one level down in the hierarchy.

**Review by fbruton — CHANGES_REQUESTED** [review]

File: src/main/java/emissary/core/IBaseDataObjectHelper.java [hunk 3]
```diff
@@ -260,4 +262,25 @@ public static byte[] findPreferredData(final IBaseDataObject payload, List<Strin
...
         return payload.data();
     }
+
+    /**
+     * Get the effective parameters for a given payload, considering parameters inherited from the parents
+     * 
+     * @param payload Get parameters from this IBaseDataObject
+     * @return The full set parameters including inherited ones
+     */
+    public static Map<String, Collection<Object>> getParametersWithInheritance(IBaseDataObject payload) {
```
Comment (fbruton) [inline src/main/java/emissary/core/IBaseDataObjectHelper.java:272]:
```suggestion
    public static Map<String, Collection<Object>> getCombinedTldParameters(IBaseDataObject payload) {
```

File: src/main/java/emissary/core/IBaseDataObjectHelper.java [hunk 4]
```diff
@@ -260,4 +262,25 @@ public static byte[] findPreferredData(final IBaseDataObject payload, List<Strin
...
+        IBaseDataObject parent = payload.getParent();
+        Set<String> inheritedParameters = payload.getInheritedParameters();
+        Map<String, Collection<Object>> effectiveParams = new HashMap<>(payload.getParameters());
+        if (parent != null) {
+            for (String key : parent.getParameterKeys()) {
+                if (inheritedParameters == null || inheritedParameters.contains(key)) {
```
Comment (fbruton) [inline src/main/java/emissary/core/IBaseDataObjectHelper.java:278]: Only add the TLD parameter if current `BaseDataObject` parameter's doesn't contain the key.

Comment (fbruton) [reply src/main/java/emissary/core/IBaseDataObjectHelper.java:278]: See LogFilter.java

File: src/main/java/emissary/core/BaseDataObject.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpg36) [reply src/main/java/emissary/core/BaseDataObject.java:197]: Naming conventions updated.

**Review by fbruton — CHANGES_REQUESTED** [review]

File: src/main/java/emissary/core/BaseDataObject.java [hunk 5]
```diff
@@ -288,6 +292,28 @@ public BaseDataObject(final byte[] newData, final String name, final String form
         }
     }
 
+    public BaseDataObject(final byte[] newData, final String name, @Nullable final String form, IBaseDataObject tld) {
+        setData(newData);
```
Comment (dev-mlb) [inline src/main/java/emissary/core/BaseDataObject.java:296]: Any reason not to use `this(newData, name, form);`? Do you not want the creation timestamp set??

**Pull request closed without merging**